databuilder/query_engines/sqlite.py

```diff
@@ -0,0 +1,178 @@
+import contextlib
+from functools import cache, cached_property, singledispatchmethod
+
+import sqlalchemy
+from sqlalchemy.dialects.sqlite.pysqlite import SQLiteDialect_pysqlite
+
+from databuilder.query_model import Function, SelectColumn, SelectPatientTable, Value
+from databuilder.sqlalchemy_utils import get_referenced_tables
+
+from .base import BaseQueryEngine
+
+
+class SQLiteQueryEngine(BaseQueryEngine):
+
+    sqlalchemy_dialect = SQLiteDialect_pysqlite
+
+    def get_query(self, column_definitions):
+        column_expressions = {
+            name: self.get_sql(definition)
+            for name, definition in column_definitions.items()
+        }
+        population_expression = column_expressions.pop("population")
+        return self.get_results_query(column_expressions, population_expression)
+
+    def get_results_query(self, column_expressions, population_expression):
+        # Get all referenced tables
+        population_tables = set(get_tables(population_expression))
+        other_tables = {
+            table for expr in column_expressions.values() for table in get_tables(expr)
+        }
+        all_tables = population_tables | other_tables
+
+        # Build a column expression for `patient_id`. In the simplest case there's a
+        # single table in the query which is guaranteed to have rows for every patient
+        # and so we can use that. But the flexibility of the query model means that we
+        # need to cope with more complex situations. Because of the outer joins between
+        # tables there may be no one table which has a row for every patient in the
+        # query.
```

Comment:
I have thoughts about population definition which I'm just writing up in a Slack thread.